os/FuseStore: fix several FuseStore issues

liewegas opened the pull request:
This fixes several issues.  Now passes updated scrub_test.py 50x.

https://github.com/ceph/ceph-qa-suite/pull/1118

tchaikov commented on qa/workunits/objectstore/test_fuse.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh -ex
+#!/bin/bash -ex
```
we removed the last bit of bashim in 15ea140b3bbf6b1bfa9eea19b0f167141fcaa2c9, why would we want to switch to bash again?

tchaikov commented on qa/workunits/objectstore/test_fuse.sh:
```diff
@@ -43,46 +43,46 @@ test -d $MNT/meta/all
-test -d $MNT/meta/all/@-1:7b3f43c4:::osd_superblock:0@/omap
-test -e $MNT/meta/all/@-1:7b3f43c4:::osd_superblock:0@/bitwise_hash
-test -e $MNT/meta/all/@-1:7b3f43c4:::osd_superblock:0@/omap_header
+mkdir $MNT/meta/all/#-1:7b3f43c4:::osd_superblock:0#
```
just a note for future readers, we started to use `#` since 4bd31faee0b0de908069287fa14c74dc889b41e5.

tchaikov commented on src/os/FuseStore.cc:
```diff
@@ -772,63 +823,86 @@ static int os_create(const char *path, mode_t mode, struct fuse_file_info *fi)
+  FuseStore *fs = static_cast<FuseStore*>(fc->private_data);
+  std::lock_guard<std::mutex> l(fs->lock);
+  FuseStore::OpenFile *o = reinterpret_cast<FuseStore::OpenFile*>(fi->fh);
+  if (--o->ref) {
```
should be `if (--o->ref == 0)`, i think.

tchaikov commented on src/os/FuseStore.cc:
```diff
@@ -594,59 +623,77 @@ static int os_open(const char *path, struct fuse_file_info *fi)
-      int r = fs->store->read(cid, oid, 0, 0, *pbl);
+      int r = fs->open_file(
+	path, fi,
+	[&](bufferlist *pbl) {
```
maybe we should pass the local variable `bpl` into `open_file()`, because if we go this way, `pbl` will still be `0` at https://github.com/ceph/ceph/pull/10723/commits/604f6b1f494f23d1964e3951ec9169faed5f1eac#diff-fc25a90da304d483f385da5c44750423R686, and hence the returned `fi` will be empty. so the following `os_read()` will run into a null `fi->fh` .

liewegas replied:
open_file fills in fi->fh.. so that ending block isn't used/needed for the object_data case

tchaikov commented:
@liewegas ping?

liewegas commented:
Sorry, got busy with the bluestore stuff.  Will get back to this 
this week!

liewegas commented:
Rebased, fixed some header issues (there was some cleanup in master), switched back to sh, and fixed the ref count issue.  Passing the tests.

Will reschedule the teuthology tests again once it builds...

xiexingguo commented:
retest this please

liewegas commented:
@tchaikov http://pulpito.ceph.com/sage-scrub-c/

tchaikov commented:
@liewegas i will review the change again tmr. and will merge it if it's good.
